src/neoxp/Node/NodeUtility.cs

```diff
@@ -208,6 +211,197 @@ public static void SignOracleResponseTransaction(ProtocolSettings settings, Expr
+            }
+
+            if (!TransactionExecutor.TryParseRpcUri(rpcUri, out var uri))
+            {
+                throw new ArgumentException($"Invalid RpcUri value \"{rpcUri}\"");
+            }
+
+            using var rpcClient = new RpcClient(uri);
+            var stateAPI = new StateAPI(rpcClient);
+
+            uint height = stateHeight;
+            if (height == 0)
+            {
+                uint? localRootIndex;
+                try
+                {
+                    (localRootIndex, _) = await stateAPI.GetStateHeightAsync();
+                }
+                catch (RpcException e)
+                {
+                    if (e.Message.Contains("Method not found"))
+                    {
+                        throw new Exception(
+                            "Could not get state information. Make sure the remote RPC server has state service support");
+                    }
+                    throw;
+                }
+
+                height = localRootIndex.HasValue ? localRootIndex.Value
+                    : throw new Exception($"Null \"{nameof(localRootIndex)}\" in state height response");
+            }
+
+            var stateRoot = await stateAPI.GetStateRootAsync(height);
+
+            // rpcClient.GetContractStateAsync returns the current ContractState, but this method needs
+            // the ContractState as it was at stateHeight. ContractManagement stores ContractState by
+            // contractHash with the prefix 8. The following code uses stateAPI.GetStateAsync to retrieve
+            // the value with that key at the height state root and then deserializes it into a ContractState
+            // instance via GetInteroperable.
+
+            var key = new byte[21];
+            key[0] = 8; // ContractManagement.Prefix_Contract
+            _contractHash.ToArray().CopyTo(key, 1);
+
+            var contractStateBuffer = await stateAPI.GetStateAsync(
+                stateRoot.RootHash, NativeContract.ContractManagement.Hash, key);
+            var contractState = new StorageItem(contractStateBuffer).GetInteroperable<ContractState>();
+            var states = await rpcClient.ExpressFindStatesAsync(stateRoot.RootHash, _contractHash, default);
+
+            return (contractState, states);
+        }
+
+        public static int PersistContract(NeoSystem neoSystem, ContractState state,
+            IReadOnlyList<(string key, string value)> storagePairs, ContractCommand.OverwriteForce force)
+        {
+            var (overwriteContract, overwriteStorage) = force switch 
+            {
+                ContractCommand.OverwriteForce.All => (true, true),
+                ContractCommand.OverwriteForce.ContractOnly => (true, false),
+                ContractCommand.OverwriteForce.None => (false, false),
+                ContractCommand.OverwriteForce.StorageOnly => (false, false),
+                _ => throw new NotSupportedException($"Invalid OverwriteForce value {force}"),
+            };
+
```

Comment:
Note, I pulled GetSnapshot into PersistContract so that the snapshot lifetime can be managed with a using statement